app/services/discourse_reactions/reaction_like_synchronizer.rb

```diff
@@ -0,0 +1,439 @@
+      )
+
+      # We want this to be all-or-nothing because the scope of each update/insert/delete
+      # batch is so dependent on the `all_affected_post_action_ids` or previously affected
+      # IDs. If we don't do this in a transaction, we could end up with a partial update
+      # on error and have no easy way of correcting data.
+      ActiveRecord::Base.transaction do
+        inserted_post_action_ids = create_missing_post_actions(excluded_from_like)
+        recovered_post_action_ids = recover_trashed_post_actions(excluded_from_like)
+
+        created_or_recovered_post_action_ids =
+          (recovered_post_action_ids + inserted_post_action_ids).uniq
+
+        Rails.logger.info(
+          "[ReactionLikeSynchronizer] Inserted #{inserted_post_action_ids.length} post action likes, recovered #{recovered_post_action_ids.length} trashed post action likes. (#{created_or_recovered_post_action_ids.length} total)",
+        )
+
+        create_missing_user_actions(created_or_recovered_post_action_ids)
+
+        trashed_post_action_ids = trash_excluded_post_actions(excluded_from_like)
+        Rails.logger.info(
+          "[ReactionLikeSynchronizer] Trashed #{trashed_post_action_ids.length} post action likes.",
+        )
+        delete_excluded_user_actions(trashed_post_action_ids)
+
+        all_affected_post_action_ids =
+          (created_or_recovered_post_action_ids + trashed_post_action_ids).uniq
+
+        update_post_like_counts(all_affected_post_action_ids)
+        update_topic_like_counts(all_affected_post_action_ids)
+        update_user_stats(all_affected_post_action_ids)
+        upsert_given_daily_likes(all_affected_post_action_ids)
+
+        TopicUser.update_post_action_cache(
+          post_id:
+            PostAction.with_deleted.where(id: all_affected_post_action_ids).pluck(:post_id).uniq,
+        )
+      end
+
+      post_run_report =
+        DiscourseReactions::MigrationReport.run(
+          print_report: false,
+          previous_report_data: pre_run_report,
+        )
+
+      Rails.logger.info(
+        "[ReactionLikeSynchronizer] Sync completed! Post-run report:\n\n#{DiscourseReactions::MigrationReport.humanized_report_data(post_run_report, pre_run_report)}",
+      )
+
+      { pre_run_report:, post_run_report: }
+    end
+
+    # Find all ReactionUser records that do not have a corresponding
+    # PostAction like record, for any reactions that are not in
+    # excluded_from_like, and create a PostAction record for each.
+    def self.create_missing_post_actions(excluded_from_like)
+      sql_query = <<~SQL
+        INSERT INTO post_actions(
+          post_id, user_id, post_action_type_id, created_at, updated_at
+        )
+        SELECT ru.post_id, ru.user_id, :pa_like, ru.created_at, ru.updated_at
+        FROM discourse_reactions_reaction_users ru
+        INNER JOIN discourse_reactions_reactions
+          ON discourse_reactions_reactions.id = ru.reaction_id
```

Comment:
Why do we need that inner join for `discourse_reactions_reactions`? To ensure that reaction is still available?

If yes, maybe we shouldn't do that. If someone reacted in the past, when a reaction was available, it should **count** even if now configuration of the forum changed as it was a valid "like".

Reply:
Sorry I am not following here. `discourse_reaction_reactions` is an aggregate table, it only stores the distinct reaction values and count for a specific post:

```
# Table name: discourse_reactions_reactions
#
#  id                   :bigint           not null, primary key
#  post_id              :integer
#  reaction_type        :integer
#  reaction_value       :string
#  reaction_users_count :integer
#  created_at           :datetime         not null
#  updated_at           :datetime         not null
```

I don't think we can assume intent on past reactions, we have to only rely on the current state of `excluded_from_like` . Maybe I am misinterpreting your meaning?

Reply:
You are right, I read it incorrectly